client_code/persistence.py

```diff
@@ -0,0 +1,166 @@
+            value
+        """
+        self._linked_column = linked_column
+        self._linked_attr = linked_attr
+
+    def __set_name__(self, owner, name):
+        if name == self._linked_column:
+            raise ValueError(
+                "Attribute name cannot be the same as the linked column name"
+            )
+        self._name = name
+
+    def __get__(self, instance, objtype=None):
+        if instance is None:
+            return self
+
+        if instance._delta:
+            return instance._delta[self._name]
+
+        if not instance._store:
+            return None
+
+        return instance._store[self._linked_column][self._linked_attr]
+
+    def __set__(self, instance, value):
+        instance._delta[self._name] = value
+
+
+class LinkedClass:
+    "A descriptor class for adding objects based on linked tables as attributes"
+
+    def __init__(self, cls, *args, linked_column=None, **kwargs):
+        self._linked_column = linked_column
+        self._cls = cls
+        self._args = args or []
+        self._kwargs = kwargs or {}
+
+    def __get__(self, instance, objtype=None):
+        if instance is None:
+            return self
+
+        return self._cls(
+            instance._store[self._linked_column], *self._args, **self._kwargs
+        )
+
+    def __set__(self, instance, value):
+        raise AttributeError(
+            "Linked Class instance is already set and cannot be changed"
+        )
+
+
+class PersistedClass:
+    _cache = {}
+    key = None
+
+    @classmethod
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        cls._snake_name = _snakify(cls.__name__)
+        for attr, value in cls.__dict__.items():
+            try:
+                is_persisted_class = issubclass(value, PersistedClass)
+            except TypeError:
+                is_persisted_class = False
+
+            if is_persisted_class:
+                setattr(cls, attr, LinkedClass(cls=value, linked_column=attr))
+
+    @classmethod
+    def search(cls, lazy=False, *args, **kwargs):
+        rows = anvil.server.call(f"search_{cls.__name__.lower()}", *args, **kwargs)
+        if lazy:
+            return (cls(store=row) for row in rows)
+
+        result = [cls(store=row) for row in rows]
+        cls.cache.clear()
+        for obj in result:
+            cls._cache[getattr(obj, cls.key)] = obj
+        return result
+
+    @classmethod
+    def get(cls, key):
+        try:
+            return cls._cache[key]
+        except KeyError:
+            row = anvil.server.call(f"get_{cls._snake_name}", {cls.key: key})
+            obj = cls(store=row)
+            cls._cache[key] = obj
+            return obj
+
+    def __init__(self, store=None, *args, **kwargs):
+        self._store = store or {}
+        self._delta = {}
+
+    def __getattr__(self, key):
+        if self._delta and key in self._delta:
+            return self._delta[key]
+
+        return dict(self._store).get(key, None)
```

Comment:
this call to `dict` seems unnecessary and we seem to be losing any AttributeErrors for missing keys
if you decide to use `q.fetch_only` in your `get` then you've just negated that here

Reply:
Yep, that's a throw back to a previous attempt which I'd forgotten all about. I'll get rid...